sl/symtrace.cc

```diff
@@ -83,7 +81,39 @@ bool hasDupChildren(const Node *node)
 
 Node::~Node()
 {
+    if (!alive_)
+        // this node is already being destroyed
+        return;
+
     alive_ = false;
+
+    // visit all parents recursively and propagate death notification
+    Node *node;
+    WorkList<Node *> wl(this);
+    while (wl.next(node)) {
+        BOOST_FOREACH(Node *parent, node->parents()) {
```

Comment:
The project uses C++11 or newer so range-based for loop would be better than `BOOST_FOREACH`.